### ISSUE TITLE
Using with MS/TP Router

### ISSUE THREAD

**Issue by mwitt1337** [BiancoRoyal/node-bacstack#28]
Node Version: `10.15.0`

Node BACstack Version: `0.2.4`

- Question

**Note:** Make sure you have read the [FAQs](https://github.com/fh1ch/node-bacstack/blob/master/FAQ.md)
before logging this issue.

### Feature Request / Question

 I am trying to use node-bacstack with devices behind a MS/TP router. It is unclear to me how to use node-bacstack to accomplish this. The NET and ADR is not obvious or clear to me. I am trying to discover the devices behind device 156 and read/write their properties. I'm new to MS/TP.

Below is the WHOIS return from /examples/read-device.js
![image](https://user-images.githubusercontent.com/26658718/88098811-5c12e180-cb68-11ea-8955-f09ba572d63f.png)

Device 156 is the MS/TP router that has child devices/properties. The readproperties function included in the /examples/read-devices.js retuned this JSON for device 156
![image](https://user-images.githubusercontent.com/26658718/88098989-95e3e800-cb68-11ea-8e0f-e6acaa83ffe6.png)

A bacnet explorer returned this for device 156
![image](https://user-images.githubusercontent.com/26658718/88099058-ab591200-cb68-11ea-9665-528694df0c09.png)

Here is code. Mostly unchanged from the example.

[read-device.zip](https://github.com/BiancoRoyal/node-bacstack/files/4955839/read-device.zip)

**Comment by Apollon77** [comment]
Can you please run your code with debug?

DEBUG=bacnet:* node yourfile.js

And post the log here

**Comment by NubeDev** [comment]
@mwitt1337  
I had the same issue and found there is support for passing in the network number and mac addr.

For me BACnet router details
`ip: 192.168.15.137`
`network num: 3`
For me BACnet device details
`BACnet device ms/tp mac: 6`

Working code example that was tested a few months ago

```javascript
// different ways to do a who is including a network whoIs
bacnetClient.whoIs();
bacnetClient.whoIs({'net':0xFFFF});
bacnetClient.whoIs({'net':0xFFFF,'interface':'0.0.0.0'});
bacnetClient.whoIs({'net':3});

//  read a point through a router to an MSTP device
bacnetClient.readProperty({address: '192.168.15.137',net:3,adr:[6]} , {type: 1, instance: 1}, 85, (err, value) => {
  console.log('value: ', value);
  console.log('err: ', err);
});

```

**Comment by Apollon77** [comment]
In theory when you use on('iAm') to get discovered devices then the net/adr should be included in the "found device details".

@mwitt1337 please try to send the "whoIs" as above ... and then we can see in the log output if the found evices include the details as needed.

Then we can enhance the docs and example

**Comment by mwitt1337** [comment]
So here are the commands I ran and the debug log output

**bacnetClient.whoIs();**

sent whoIs 1595634894063
examples/read-device.js:468
Found Device 3001 on {"address":"10.10.0.22","forwardedFrom":null}
examples/read-device.js:490
Found Device 1234 on {"address":"10.10.0.38","forwardedFrom":null}
examples/read-device.js:490
Found Device 9101 on {"address":"10.10.0.47","forwardedFrom":null}
examples/read-device.js:490
Found Device 156 on {"address":"10.10.0.48","forwardedFrom":null}

**bacnetClient.whoIs({'net':0xFFFF});**

sent whoIs 1595634940669
examples/read-device.js:468
Found Device 3001 on {"address":"10.10.0.22","forwardedFrom":null}
examples/read-device.js:490
Found Device 1234 on {"address":"10.10.0.38","forwardedFrom":null}
examples/read-device.js:490
Found Device 9101 on {"address":"10.10.0.47","forwardedFrom":null}
examples/read-device.js:490
Found Device 156 on {"address":"10.10.0.48","forwardedFrom":null}
examples/read-device.js:490
Found Device 985911 on {"address":"10.10.0.48","forwardedFrom":null}
examples/read-device.js:490
Found Device 862527 on {"address":"10.10.0.48","forwardedFrom":null}
examples/read-device.js:490
Found Device 48871 on {"address":"10.10.0.47","forwardedFrom":null}


**bacnetClient.whoIs({'net':0xFFFF,'interface':'10.10.0.48'});**

sent whoIs 1595635017212
examples/read-device.js:468
Found Device 3001 on {"address":"10.10.0.22","forwardedFrom":null}
examples/read-device.js:490
Found Device 156 on {"address":"10.10.0.48","forwardedFrom":null}
examples/read-device.js:490
Found Device 1234 on {"address":"10.10.0.38","forwardedFrom":null}
examples/read-device.js:490
Found Device 9101 on {"address":"10.10.0.47","forwardedFrom":null}
examples/read-device.js:490
Found Device 48871 on {"address":"10.10.0.47","forwardedFrom":null}
examples/read-device.js:490
Found Device 985911 on {"address":"10.10.0.48","forwardedFrom":null}
examples/read-device.js:490
Found Device 862527 on {"address":"10.10.0.48","forwardedFrom":null}

**bacnetClient.whoIs({'net':3});**

sent whoIs 1595635055268

**bacnetClient.whoIs({'net':0});**

sent whoIs 1595635116749
examples/read-device.js:468
Found Device 3001 on {"address":"10.10.0.22","forwardedFrom":null}
examples/read-device.js:490
Found Device 156 on {"address":"10.10.0.48","forwardedFrom":null}
examples/read-device.js:490
Found Device 9101 on {"address":"10.10.0.47","forwardedFrom":null}
examples/read-device.js:490
Found Device 1234 on {"address":"10.10.0.38","forwardedFrom":null}


Debug isn't as much information as I expected it to. Device 156 is the MS/TP route and I need to access the devices attached to it.

**Comment by Apollon77** [comment]
Yes the debug should be waaaaay more. Please try against and use

DEBUG=bacnet* node Myfile.js

Please

**Comment by mwitt1337** [comment]
Sorry, I'm still newer to node and I'm using VSCode. What I posted
previously is from the debug console. Is there a logfile I should be
looking at?
Running this in a javascript debug terminal returns errors DEBUG=bacnet* is
not recognized as the name of a cmdlet. I'm missing something fundamental
here.
Running DEBUG=Bacnet*; node read-device.js works but doesn't return
any further detail then what I've posted before.

**Comment by Apollon77** [comment]
Just open a shell and "cd" in the relevant directory and run it from there as described above

**Comment by mwitt1337** [comment]
Thanks for being patient with me ;) Got the debug to work. I loaded the
source into a folder but didn't run npm install...

[image: image.png]

**Comment by biancode** [comment]
Hi, read some more here about debug https://github.com/BiancoRoyal/node-red-contrib-modbus/wiki/DEBUG

But with bacnet ;)

Like:


on CMD:

set DEBUG=bacnet*

on PowerShell:

$env:DEBUG = "bacnet*"

**Comment by mwitt1337** [comment]
The image from my email response didn't post. Here is the debug output. It really doesn't seem to have much more information. Does DEBUG=bacnet* already include all of the options specified in the link?

![image](https://user-images.githubusercontent.com/26658718/88575561-13e53a80-d012-11ea-820c-bf892bcf113b.png)

**Comment by Apollon77** [comment]
which way you sent the whois command?

>  bacnetClient.whoIs({'net':0xFFFF});
bacnetClient.whoIs({'net':0xFFFF,'interface':'0.0.0.0'});
bacnetClient.whoIs({'net':3});

?

**Comment by mwitt1337** [comment]
I ran the who is command 4 different ways and attached a text file of the outputs.

[net 0.txt](https://github.com/BiancoRoyal/node-bacstack/files/4986335/net.0.txt)
[net 0xFFFF.txt](https://github.com/BiancoRoyal/node-bacstack/files/4986336/net.0xFFFF.txt)
[net 3.txt](https://github.com/BiancoRoyal/node-bacstack/files/4986337/net.3.txt)
[new 0xFFFF with interface.txt](https://github.com/BiancoRoyal/node-bacstack/files/4986338/new.0xFFFF.with.interface.txt)

**Comment by NubeDev** [comment]
Are you using wireshark to debug?
Have you tried to read a point value or anything different to a whoIs? 

Like this
```javascript
//  read a point through a router to an MSTP device
bacnetClient.readProperty({address: '192.168.15.137',net:3,adr:[6]} , {type: 1, instance: 1}, 85, (err, value) => {
  console.log('value: ', value);
  console.log('err: ', err);
});

```

**Comment by mwitt1337** [comment]
I was able to make some progress on one of the devices. 

`  
bacnetClient.on('listening', () => {
   bacnetClient.readProperty({address: '10.10.0.47',net:2002,adr:[36]} , {type: 8, instance: 48871}, 76, (err, value) => {
    printResultObject(48871,value);
    console.log('value: ', value);
    console.log('err: ', err); 
  });
`
I attached a json file with the printResultObject value. 
[resultobject.txt](https://github.com/BiancoRoyal/node-bacstack/files/5013869/resultobject.txt)

Below is a screen shot of the json from the attached file visualized. At the top the Type:8 and Instance: 48871 matches the light panel I was trying to read values from. The child objects at the bottom are confusing me because they are Type:12 which I haven't found what object type 12 is. The value is Type: 2 which is analog output and instance: 601 must be the instance of that analog value. 

I was able to read the current value of instance 601
`  
bacnetClient.on('listening', () => {
   bacnetClient.readProperty({address: '10.10.0.47',net:2002,adr:[36]} , {type: 2, instance: 601}, 85, (err, value) => {
    printResultObject(601,value);
    console.log('value: ', value);
    console.log('err: ', err); 
  });
`
This is a major break through and hopefully this helps anyone else trying to sort out reading properties. Next I will be testing writing to these properties.
![image](https://user-images.githubusercontent.com/26658718/89146829-e40ec900-d522-11ea-8e6b-5a4283404b41.png)

**Comment by Apollon77** [comment]
But you were still not able to get that device by a whoIs call?

**Comment by mwitt1337** [comment]
No, the devices in the object list are not coming back with the whois. The
MSTP bridge will return with the whois but none of the child devices.

**Comment by mwitt1337** [comment]
These are buttons and switches  on a light panel so I don't know if they
should be responding to a whois

**Comment by Apollon77** [comment]
Interesting is - when checking your screenshot above - that the router device returns some more data in the response. maybe we need to look in how/that they are parsed correctly :-(
Maybe @NateZimmer have an idea? The logic should be the same as in the official repo

**Comment by NateZimmer** [comment]
**To people having the issues:** 
- Ensure you run wireshark while performing tests. Wireshark will give you a nice dissection of the who-is response / generic bacnet responses. From there you can tell if the sub devices are actually being reported. Ideally somehow upload the exported pcap filtering on the bacnet port. 

@mwitt1337 Regarding this example: 
![image](https://user-images.githubusercontent.com/1478536/89470474-7ab8d100-d741-11ea-8c15-e116625b845d.png)
Here the network number appears to be 1
Here the network address appears to be a 6 byte MAC address. I take it, this is not a MS/TP device then? MS/TP devices will have a network address of a single byte. TBH I haven't tried/tested on my personal fork of addressing by 6byte MAC address. I wonder if whatever bacnet explorer you are using might be masking the network address though its possible it could be the MAC address. Wireshark will tell you the answer  

You should first: 
1. Send a who-is to the bacnet router 
2. Create an iAm event listener and log the iAm responses you get 

From there you will know how to address the sub devices 

**To this repo's maintainers:**
What did you do with the iAm event emitter? https://github.com/fh1ch/node-bacstack/blob/master/lib/client.js#L290 . 

The iAm event emitter should contain:

- Address:
-- Router IP
-- Network number
-- Network address 
- Other info (device ID, max apdu, ...etc)
